### PR TITLE
Check if new VM name is correct when renaming

### LINKED ISSUE #3137: 'virtctl rename' accepts invalid names
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
When providing an invalid name to `virtctl rename` the name is accepted:
`virtctl rename vm-cirros "a b"`

**What you expected to happen**:
An error should be returned to the user and the renaming process should not initiate, instead the error is written to `virt-controller` logs and the controller will attempt to rename the VM forever, and will only stop once the VM is deleted.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a VM
2. Rename it as follows:
`virtctl rename [vm-name] "a b"

**Anything else we need to know?**:
Proposed solution:
In `virt-api`, check if the new name of a VM is a valid name by using this validation function:
https://github.com/kubernetes/kubernetes/blob/5383e01d2aba769dcf89c3b9c5092a6432193b61/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L44

### PR DESCRIPTION
**Pull request by akrejcir**
**What this PR does / why we need it**:
Adds name validation for `virtctl rename`. Now only valid kubernetes identifiers can be used.

**Which issue(s) this PR fixes**:
Fixes #3137

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-api/rest/subresource.go [hunk 1]
```diff
@@ -560,6 +561,13 @@ func (app *SubresourceAPIApp) RenameVMRequestHandler(request *restful.Request, r
 		return
 	}
 
+	nameErrs := k8sapivalidation.NameIsDNSSubdomain(opts.NewName, false)
+	if len(nameErrs) > 0 {
+		message := fmt.Sprintf("The VM's new name is not valid: %s", strings.Join(nameErrs, "; "))
```
Comment (davidvossel) [inline pkg/virt-api/rest/subresource.go]: We should put this validation check server side in the `pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go` instead.

That will still cause the request to rename using virtctl to fail, but it will be validated on the server side with the rest of the spec/status validators instead.  So, the request to PATCH will fail here in the client if the validation webhook rejects the change.

File: staging/src/kubevirt.io/client-go/api/v1/types.go [hunk 2]
```diff
@@ -946,7 +946,7 @@ const (
 	VirtualMachinePaused VirtualMachineConditionType = "Paused"
 
 	// This condition indicates that the VM was renamed
-	RenameConditionType VirtualMachineConditionType = "Rename Operation"
```
Comment (davidvossel) [inline staging/src/kubevirt.io/client-go/api/v1/types.go]: ouch, I missed this one.  We're not technically supposed to change things like this because of compatibility, but maybe if there hasn't already been a kubevirt release since this was introduced we'll be okay.

File: pkg/virt-api/rest/subresource.go (hunk 1 above)
Comment (akrejcir) [reply pkg/virt-api/rest/subresource.go]: Done.

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 2 above)
Comment (akrejcir) [reply staging/src/kubevirt.io/client-go/api/v1/types.go]: The PR that introduced this was merged after the v0.27.0 tag. So it hasn't been released yet.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 3]
```diff
@@ -238,62 +231,78 @@ func ValidateVirtualMachineSpec(field *k8sfield.Path, spec *v1.VirtualMachineSpe
...
-			}
-		}
+	if renameRequest == nil {
+		return nil
 	}
 
-	return causes
+	// Reject rename requests if the VM is running
+	runningStatus, _ := vm.RunStrategy()
+	if runningStatus != v1.RunStrategyHalted {
```
Comment (danielBelenky) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: If running status is manual, the VMI can be stopped by the user manually... shouldn't we allow rename in that case?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 4]
```diff
@@ -238,62 +231,78 @@ func ValidateVirtualMachineSpec(field *k8sfield.Path, spec *v1.VirtualMachineSpe
...
+		return []metav1.StatusCause{{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: "Cannot rename a running VM",
+			Field:   k8sfield.NewPath("spec", "running").String(),
+		}}
+	}
+
+	newName, hasNewName := renameRequest.Data["newName"]
+	if !hasNewName || newName == "" {
```
Comment (danielBelenky) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: checking that `newName` is empty here is wrong.
Since a user can set an empty string as a new name and the error in result should be that this is not a legal name not that the name was not provided.

"" is not a DNS subdomain since DNS subdomain start and end with an alphanumeric character.

File: pkg/virt-controller/watch/vm.go [hunk 5]
```diff
@@ -324,7 +324,7 @@ func (c *VMController) handleVMRenameRequest(vm *virtv1.VirtualMachine, newName
 	newVM.Status.Conditions = []virtv1.VirtualMachineCondition{
 		{
 			Type:    virtv1.RenameConditionType,
-			Status:  "Success",
+			Status:  k8score.ConditionTrue,
```
Comment (danielBelenky) [inline pkg/virt-controller/watch/vm.go]: Please try to avoid unrelated changes in PRs.. even if they are very small (unless they are required for the functionality of this PR)

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go (hunk 3 above)
Comment (akrejcir) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: Renaming is done by creating a new VM object and removing the old one. So renaming a running VM would stop the VM and remove the VMI object. That is probably not what we want.
But I see that this code is not correct. If the strategy is 'manual', it would not allow rename even if VM is not running.

Comment (danielBelenky) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: > But I see that this code is not correct. If the strategy is 'manual', it would not allow rename even if VM is not running.
Yep.. this is what I meant.

File: staging/src/kubevirt.io/client-go/api/v1/types.go (hunk 2 above)
Comment (davidvossel) [reply staging/src/kubevirt.io/client-go/api/v1/types.go]: yay!

File: pkg/virt-controller/watch/vm.go (hunk 5 above)
Comment (akrejcir) [reply pkg/virt-controller/watch/vm.go]: Ok, I will post another PR.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go (hunk 4 above)
Comment (akrejcir) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: Done

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go (hunk 3 above)
Comment (akrejcir) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: Done
